src/routers/alpha-router/alpha-router.ts

```diff
@@ -119,51 +119,85 @@ export class AlphaRouter implements IRouter<AlphaRouterConfig>, ISwapToRatio<Alp
-        sqrtPriceX96Upper,
-        precision,
-        true
-      )
-      const token1Proportion = SqrtPriceMath.getAmount1Delta(
-        sqrtPriceX96,
-        sqrtPriceX96Lower,
-        precision,
-        true
+      let optimalRatio = this.calculateOptimalRatio(
+        position,
+        position.pool.sqrtRatioX96
       )
 
-      const amountToSwap = calculateRatioAmountIn(
-        new Fraction(token0Proportion, token1Proportion),
-        position.pool.token0Price,
-        token0Balance,
-        token1Balance,
-      )
+      let zeroForOne: boolean
+      if (position.pool.tickCurrent < position.tickLower) {
+        zeroForOne = true
+      } else if (position.pool.tickCurrent > position.tickUpper) {
+        zeroForOne = false
+      } else {
+        zeroForOne = new Fraction(token0Balance.quotient, token1Balance.quotient).greaterThan(optimalRatio)
+        if (!zeroForOne) optimalRatio = optimalRatio.invert()
+      }
 
-      return this.routeExactIn(
-        amountToSwap.currency,
-        amountToSwap.currency == token0Balance.currency
-          ? token1Balance.currency
-          : token0Balance.currency,
-        amountToSwap,
-        swapConfig,
-        routingConfig
-      )
+      const [inputBalance, outputBalance] = zeroForOne
+        ? [token0Balance, token1Balance]
+        : [token1Balance, token0Balance]
+
+      let exchangeRate: Fraction = zeroForOne
+        ? position.pool.token0Price
+        : position.pool.token1Price
+      let swap: SwapRoute<TradeType.EXACT_INPUT> | null = null
+      let ratioAchieved = false
+
+      while (!ratioAchieved) {
+        let amountToSwap = calculateRatioAmountIn(
+          optimalRatio,
+          exchangeRate,
+          inputBalance,
+          outputBalance,
+        )
+
+        swap = await this.routeExactIn(
+          inputBalance.currency,
+          outputBalance.currency,
+          amountToSwap,
+          swapConfig,
+          routingConfig
+        )
+        if (!swap) {
+          throw new Error('did not fetch swap')
+        }
+
+        let inputBalanceUpdated = inputBalance.subtract(swap.trade.inputAmount)
+        let outputBalanceUpdated = outputBalance.add(swap.trade.outputAmount)
+        let newRatio = inputBalanceUpdated.divide(outputBalanceUpdated)
+
+        swap.route.forEach(route => {
+          route.route.pools.forEach((pool, i) => {
+            if(
+              pool.token0 == position.pool.token0 &&
+              pool.token1 == position.pool.token1 &&
+              pool.fee == position.pool.fee
+            ) {
+              optimalRatio = this.calculateOptimalRatio(
+                position,
+                JSBI.BigInt(route.sqrtPriceX96AfterList[i]!.toString())
+              )
+            }
+          })
+        })
+
+        ratioAchieved = (
+          newRatio.equalTo(optimalRatio) ||
+          this.absoluteValue(newRatio.asFraction.divide(optimalRatio).subtract(1)).lessThan(errorTolerance)
+        )
```

Comment:
Do we have any sense of how many iterations it will take to converge? And if we are guaranteed to converge? 

Wondering 
1/ if it would be possible that we keep flipping between a ratio that is below our optimal and a ratio above our optimal causing an inf loop
2/ whether we should put a 'max iterations' to prevent infinite loops or the calculation taking too long / consuming too many resources
3/ whether we should terminate as soon as we are within the error threshold or not- continuing iterating could find us a swap that results in us being even closer to the optimal ratio and the user having less dust at the end

Reply:
When I do giant trades trading 200 million dollar values, I converge at about 4 trades. I think I still need to do some larger scoped tests against many different tokens/scenarios. 
1) investigating infinite loops I think is really important. Other than thinking hard through some potential scenarios, the best way I can think of is to try this with many different pools with many different amounts, but this is something I'm open to suggestions on.
2) this is in my todo, in the PR description unchecked. will add that,
3) I was going with the former, but open to ideas.

Reply:
1. I'm not sure, but I think having max iterations will go a long way here
2. Ah ok great, sorry didnt see that
3. We could have something like a minimum iterations too. It could be tuned to give us a latency we're happy with. Feels like something we can follow up with though